### DISCUSSION TITLE
Local Integration Testing

### DISCUSSION THREAD

**Discussion by amerryma** [supabase/supabase#16415] [Questions]
Hi, I'm working through a strategy for integration testing of a local Supabase instance. My integration suite is written with Jest and it uses the node pg driver to connect to the database and run assertions. It takes care of seeding and resetting the database. Right now my tests essentially use the Supabase CLI to start Supabase.

My problem comes from the fact that the Supabase CLI can only interact with a single database at a time, so if I want to run integration tests, it will affect my local development environment. I want to explore options where my integration tests don't need to worry about dealing with a shared database.

Options I've considered:

1. Run Supabase within a docker container (docker in docker). This way I can run multiple Supabases on the same machine pointing to the same local directory. Not sure if this is entirely feasible.
2. Create a different directory for supabase-cli to run in, and modify the config.toml to use different ports (symlink to share migrations, etc from the main project)
3. Just share the db between integration tests and local dev. All data generated with the integration tests should be tracked (prefix or something) so that I can delete them after the tests are done.
4. Migrate everything to pgTap. This really isn't what I want to do as I like testing the supabase system in it's entirety using the supabase js client. I'm focused on full integration testing not just unit testing the database.

I've seen the discussions [here](https://github.com/orgs/supabase/discussions/14329) and wondering if this is still the preferred method.

**Comment by brahimmouhamou** [comment]
Hey Aaron, it has been a while, what did you actually choose and would you still recommend it?

**Reply by amerryma** [reply]
We went with option 3 for most of my Supabase projects when I want to use the supabase js client to make assertions.

Although, I've written most of my tests just connecting to the db via node postgres and running assertions within a transaction, then roll it back.

**Reply by nickdichev** [reply]
> just connecting to the db via node postgres and running assertions within a transaction

Does this mean you are doing all the DB operations outside of the supabase client and directly interfacing with a postgres client to do all reads/writes? What benefit do you get from supabase if you are not interfacing with the client?

My initial thought was that you were opening a transaction with the node pg client and then making supabase client calls, but this can't work because the supabase client won't be in the same transaction as the pg client.

**Reply by amerryma** [reply]
You can emulate what the client does by simply opening a transaction, setting the jwt (so that auth.jwt() and auth.uid() works), and making queries as a particular user. It's quite effective and incredibly efficient.

**Reply by amerryma** [reply]
supabase-js is essentially just PostgREST and the way they interact with the database is pretty straightforward, not much benefit I see in verifying that PostgREST is working correctly. I mainly want to test the business logic that I'm programming into RLS and RPC methods.

**Reply by nickdichev** [reply]
Thanks for the response, that makes sense for what you're testing. I actually do want to test through the postgrest boundary.

I'm trying to emulate a feature of how Elixir app do unit tests which is having an isolated DB state per-test.

**Reply by amerryma** [reply]
So are you using the Supabase JS client inside of your backend (Elixir)? I only ever used the Supabase JS client from the front-end (React).

**Reply by nickdichev** [reply]
No, I've been doing web apps in elixir for like 5 years and now I'm just playing around in a supabase/react app

**Reply by hems** [reply]
> You can emulate what the client does by simply opening a transaction, setting the jwt (so that auth.jwt() and auth.uid() works), and making queries as a particular user. It's quite effective and incredibly efficient.

this is really interesting. still i believe we would benefit from completely deleting all the data on the database before tests start and then compeltely deleting it between tests / when tests end.

I'm in touch with supabase support trying to figure out how to do this.

My work rely heavily on Test Driven Development so that mean every time i save a file i run at least all the tests related to this file, which means i need a database that's not my local development to run, with the same tables, policies, etcs, but without interfering with the local database.

Also the tests should run / tear up / tear down compeltely independent of the local database.

So it seems to me it's more the case of "duplicating" the database, perhaps the test database could run all the migrations from the local dev but reset all the data, something among those lines...

i'm beginning on supabase but really invested in finding out the perfect solution for this problem.

I'm currently coming from a MongoDB solution where i can do all of that locally with absolute no problems by running [Mongo Memory Server](https://github.com/typegoose/mongodb-memory-server) which is basically a copy of the database that lives in memory and is deleted when the tests finish, when running tests all my API, application, tests, etcs, points to that database and when the tests finish the database is wiped.

Works like a charm. I would love to find a solution like that ( not necessarily with in memory database, but with the same sort of easy of use ) for supabase.

**Reply by amerryma** [reply]
Database tear-down and start-up is very expensive. Some strategies I've used will do truncates on tables between tests but still takes much longer than a simple rollback strategy. I was heavily inspired while following the testing guidelines laid out in a similar project [here](https://www.graphile.org/postgraphile/testing-jest/). For example, I have over 4k tests running with against a real database (1-5 sql statements per test) which take about 1 minute total to run.

**Reply by amerryma** [reply]
> I'm currently coming from a MongoDB solution where i can do all of that locally with absolute no problems by running [Mongo Memory Server](https://github.com/typegoose/mongodb-memory-server) which is basically a copy of the database that lives in memory and is deleted when the tests finish, when running tests all my API, application, tests, etcs, points to that database and when the tests finish the database is wiped.

I also came from a mongo memory server about 5 years ago and I was very surprised they didn't have a way to do it in the postgres ecosystem. But that space is definitely evolving because of a project called [pglite](https://github.com/electric-sql/pglite) which may end up allowing you to do something similar.

**Reply by hems** [reply]
> Database tear-down and start-up is very expensive. Some strategies I've used will do truncates on tables between tests but still takes much longer than a simple rollback strategy. I was heavily inspired while following the testing guidelines laid out in a similar project [here](https://www.graphile.org/postgraphile/testing-jest/). For example, I have over 4k tests running with against a real database (1-5 sql statements per test) which take about 1 minute total to run.

This seens interesting, but if you're testing a function that for instance requires seeding and then multiple calls, can you still rollback the end?

I really like the example where it creates a user and then becomes the user, that's really helpful.

In a way, I think the analogue solution to Memory Mongo Server would be to:

1. run a "test-project" similar to what you suggested by changing the config.toml
2. run the migrations from the local development into that, so all the tables and policies are loaded into the test project
3. between each test delete all data from all databases and leave just the tables / policies untouched
3. "reset the database" from the test project at the very end

**Comment by Slyracoon23** [comment]
I actually experimented with all the different approaches you, @amerryma, mentioned, and ended up finding the prefix method (option 3) to be the easiest and cleanest solution for my use case.

The prefix approach let me keep things simple.

The keys were:

* Prefixing all test data with something like "test_" to easily identify it.
* Creating a custom RPC function that did a cascading delete of all prefixed data after test runs.

This let me share the database between tests and local dev, while keeping things tidy. Just had to be disciplined about cleaning up after each test run.

In the end, the prefix method was the most straightforward way to get full integration tests running without too much overhead or risk of polluting my local env. Still requires a custom RPC function that needs to be modified to the size of affected tables. Also, it must maintain the order in which they reference each other.

Still best possible solution.

**Reply by devsheva** [reply]
This seems a good solution, but what if you have a method that does a simple select? How will you exclude the "non" test prefixed ones from being affected? Imagine you're doing a bulk update inside that method

**Comment by amerryma** [comment]
So, one simple option I thought of was just simply changing the project name within `config.toml`. It will give you an entirely separate database and docker images.

**Comment by devsheva** [comment]
What I did is creating a different `test` schema, so you have a completely different dataset while using the same project DB

**Reply by hems** [reply]
any links to documentation / examples of how this works?

**Reply by amerryma** [reply]
Only official documentation for alternate schemas is [here](https://supabase.com/docs/guides/api/using-custom-schemas).

**Reply by hems** [reply]
> What I did is creating a different `test` schema, so you have a completely different dataset while using the same project DB

do you have an example somewhere with the tear up / tear down / cleaning between tests code?

that would be wicked!

**Reply by devsheva** [reply]
@hems Sorry for the late answer, but here is how I did in my project.

1. Set an environment variable `APP_ENV` to test when needed, and use it like [this](https://github.com/devsheva/hn-telegram-bot/blob/master/src/utils.ts)
2. define seed and cleanup functions with this [lines](https://github.com/devsheva/hn-telegram-bot/blob/dc5047197f4830d0ccbf1633853dc8c6fc464d97/src/utils/test_helpers.ts#L133-L153)
4. use them respectively in `beforeAll` and `afterAll` hooks like [here](https://github.com/devsheva/hn-telegram-bot/blob/dc5047197f4830d0ccbf1633853dc8c6fc464d97/src/__tests__/bot/daily_analysis.spec.ts#L26-L32)

**Comment by Floriferous** [comment]
Similar story for me than @hems and @amerryma. Was used to running a in-memory mongo server for all my unit tests. Always test against the database and avoid mocking to have proper confidence in your test suite.

I don't know what the right equivalent for Supabase is. I don't want to wipe my local supabsae instance every time I run the tests, and I also want test setup to be super fast so I never hesitate to run the whole test suite because it may take a minute and break my flow.

Maybe it should be possible to connect to the local supabase in test mode, and then it just reuses the schemas/migrations, but not the actual data. Maintaining 2 different supabase configs doesn't feel right.

And of course having a good story to run those same tests in the CI would be ideal.

**Reply by hems** [reply]
@Floriferous last time i spoke with support it seemed like they were far away from a solution to be able to run tests offline for instance which is very necessary if you want to run tests fasts when developing. anyhow good luck! i have gave up on this.
